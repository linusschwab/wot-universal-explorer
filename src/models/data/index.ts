export {Data} from './Data'
export {InputData} from './InputData'
export {OutputData} from './OutputData'
