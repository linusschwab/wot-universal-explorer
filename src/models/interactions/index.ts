export {InteractionData} from './InteractionData';
export {InteractionPattern} from './InteractionPattern';
export {Action} from './Action';
export {Event} from './Event';
export {Property} from './Property';
export {ISubscriber} from './ISubscriber';
