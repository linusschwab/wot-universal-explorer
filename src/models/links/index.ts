export {Link} from './Link';
export {HTTPLink} from './HTTPLink';
export {COAPLink} from './COAPLink';
export {Operation} from './Operation';
