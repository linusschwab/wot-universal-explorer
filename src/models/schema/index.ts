export {DataSchema} from './DataSchema'
export {InputSchema} from './InputSchema'
export {OutputSchema} from './OutputSchema'
