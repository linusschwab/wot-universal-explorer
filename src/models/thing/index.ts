export {Thing} from './Thing';
export {ThingsManager} from './ThingsManager';
