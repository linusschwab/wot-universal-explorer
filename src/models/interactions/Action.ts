import {Interaction} from "./Interaction";
import {InputData, OutputData} from "../data";


export class Action extends Interaction {

    public inputData: InputData;

    constructor(name: string, outputData: OutputData, inputData: InputData) {
        super(name);
        this.outputData = outputData;
        this.inputData = inputData;
    }

    public invoke() {
        // TODO: Implement
        return true;
    }

    public toString() {
        return this.name + ' action';
    }

}
