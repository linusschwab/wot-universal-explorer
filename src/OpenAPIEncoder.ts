import {Thing} from "./models/thing/Thing";
import {HTTPLink, Link, Operation} from "./models/links";
import {DataSchema, InputSchema} from "./models/schema";
import {isArray} from "util";
import {Action} from "./models/interaction";


export class OpenAPIEncoder {

    public static encode(thing: Thing) {
        return JSON.stringify(thing, (key: string, value: Object) => {
            if (value instanceof Thing) {
                return {
                    "openapi": "3.0.0",
                    "info": this.info(value.name),
                    "paths": this.paths(value.links),
                }
            }
            // TODO: Support other link types
            if (value instanceof HTTPLink) {
                return this.operations(value);
            }
            if (value instanceof Operation) {
                return this.operation(value);
            }

            return value;
        }, 4);
    }

    private static info(name: string) {
        return {
            "version": "1.0",
            "title": name,
            "description": "API to interact with " + name + " generated from the TD",
        };
    }

    private static paths(links: Link[]) {
        let obj: any = {};

        for (let link of links) {
            obj[link.href] = link;
        }

        return obj
    }

    private static operations(link: HTTPLink) {
        let obj: any = {};

        for (let operation of link.operations) {
            obj[operation.type] = operation;
        }

        return obj;
    }

    private static operation(o: Operation) {
        let operation: any = {
            "description": o.description,
            "responses": {
                "200": {
                    "description": "success"
                }
            },
            "tags": [
                (<any>o.link.interaction).constructor.name
            ]
        };

        let requestBodyData = [];
        if (o.link.interaction instanceof Action) {
            let inputData = o.link.interaction.inputSchema;
            if (inputData) {
                requestBodyData.push(inputData);
            }
        }

        /*let outputData = o.link.interaction.schema;
        if (outputData) {
            operation['responses']['200']['content'] = {
                [o.link.mediaType]: {
                    "schema": this.schema(outputData)
                }
            };

            if (outputData.writable) {
                requestBodyData.push(outputData);
            }
        }*/

        if (requestBodyData.length > 0) {
            operation['requestBody'] = this.requestBody(o.link.interaction.toString(), o.link.mediaType, requestBodyData);
        }

        return operation;
    }

    private static schema(data: DataSchema | DataSchema[]) {
        if (!isArray(data)) {
            data = [data];
        }

        // TODO: Check if same name multiple times
        let properties: any = {};
        let required: string[] = [];

        for (let d of data) {
            properties[d.type] = {
                "type": d.type
            };

            if (d instanceof InputSchema) {
                required.push(d.type);
            }
        }

        let schema: any = {
            "type": "object",
            "properties": properties
        };

        if (required.length > 0) {
            schema['required'] = required;
        }

        return schema;
    }

    private static requestBody(description: string, mediaType: string, data: DataSchema | DataSchema[]) {
        return {
            "description": description + " parameters",
            "required": "true",
            "content": {
                [mediaType]: {
                    "schema": this.schema(data)
                }
            }
        }
    }

}
