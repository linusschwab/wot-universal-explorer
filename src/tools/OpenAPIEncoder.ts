import {Thing, ThingsManager} from "../models/thing";
import {Operation} from "../models/links";
import {DataSchema, InputSchema} from "../models/schema";
import {Action, InteractionPattern, Property, Event} from "../models/interaction";
import {isArray} from "util";


export class OpenAPIEncoder {

    public static encode(things: ThingsManager) {
        return JSON.stringify(things, (key: string, value: Object) => {
            if (value instanceof ThingsManager) {
                return {
                    "openapi": "3.0.0",
                    "info": this.info(),
                    "servers": this.servers(),
                    "paths": this.paths(value),
                }
            }
            if (value instanceof InteractionPattern) {
                return this.operations(value);
            }
            if (value instanceof Operation) {
                return this.operation(value);
            }

            return value;
        }, 4);
    }

    private static info() {
        return {
            "version": "1.0",
            "title": "Universal Explorer for the Web of Things",
            "description": "API to interact with things",
        };
    }

    private static servers() {
        // TODO: Remove hardcoded url, read from config
        return [
            {
                "url": "http://localhost:5000/things",
                "description": "Development server"
            }
        ]
    }

    private static paths(things: ThingsManager) {
        let obj: any = {};

        for (let interaction of things.getInteractions()) {
            obj[interaction.url] = interaction;
        }

        return obj
    }

    private static operations(interaction: InteractionPattern) {
        let obj: any = {};

        for (let operation of interaction.operations) {
            obj[operation.type] = operation;
        }

        return obj;
    }

    private static operation(o: Operation) {
        let operation: any = {
            "description": o.description,
            "responses": {
                "200": {
                    "description": "success"
                }
            },
            "tags": [
                o.interaction.thing.name
            ]
        };

        // Request body
        let requestSchema: DataSchema = this.requestSchema(o);
        if (requestSchema) {
            operation['requestBody'] = this.requestBody(
                o.interaction.toString(),
                'application/json',
                requestSchema
            );
        }

        // Response
        let responseSchema: DataSchema = this.responseSchema(o);
        if (responseSchema) {
            operation['responses']['200']['content'] = {
                ['application/json']: {
                    "schema": this.schema(responseSchema)
                }
            };
        }

        return operation;
    }

    private static requestSchema(o: Operation): DataSchema {
        let requestSchema: DataSchema = null;

        // Check if operation type allows request body
        if (o.type !== 'post' && o.type !== 'put') {
            return null;
        }

        // Transform interaction data schema
        if (o.interaction instanceof Property) {
            if (o.interaction.schema && o.interaction.schema.writable) {
                requestSchema = o.interaction.schema;
            }
        }
        else if (o.interaction instanceof Action) {
            requestSchema = o.interaction.inputSchema;
        }
        else if (o.interaction instanceof Event) {
            requestSchema = null;
        }

        return requestSchema
    }

    private static responseSchema(o: Operation): DataSchema {
        if (o.interaction instanceof Property) {
            return o.interaction.schema;
        }
        else if (o.interaction instanceof Action) {
            return o.interaction.outputSchema;
        }
        else if (o.interaction instanceof Event) {
            return o.interaction.schema;
        }
        else {
            return null;
        }
    }

    private static schema(data: DataSchema | DataSchema[]) {
        if (!isArray(data)) {
            data = [data];
        }

        // TODO: Check if same name multiple times
        let properties: any = {};
        let required: string[] = [];

        for (let d of data) {
            properties[d.type] = {
                "type": d.type
            };

            if (d instanceof InputSchema) {
                required.push(d.type);
            }
        }

        let schema: any = {
            "type": "object",
            "properties": properties
        };

        if (required.length > 0) {
            schema['required'] = required;
        }

        return schema;
    }

    private static requestBody(description: string, mediaType: string, data: DataSchema | DataSchema[]) {
        return {
            "description": description + " parameters",
            "required": "true",
            "content": {
                [mediaType]: {
                    "schema": this.schema(data)
                }
            }
        }
    }

}
