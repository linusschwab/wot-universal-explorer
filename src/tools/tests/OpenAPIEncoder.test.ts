import {Action, Property} from "../../models/interactions";
import {DataSchema, InputSchema, OutputSchema} from "../../models/schema";
import {HTTPLink} from "../../models/links";
import {Thing, ThingsManager} from "../../models/thing";
import {OpenAPIEncoder} from "../OpenAPIEncoder";


test('encode one thing', () => {
    // Prepare test thing
    const thing = new Thing('Testthing', 'Thing', 'http://localhost');

    const property = new Property('testProperty', new DataSchema('integer', true), true, false);
    property.registerLink(new HTTPLink('/testproperty', thing.base, 'application/json'));
    thing.registerInteraction(property);

    const action = new Action('testAction', new InputSchema('boolean'), new OutputSchema('boolean'));
    action.registerLink(new HTTPLink('/testaction', thing.base, 'application/json'));
    thing.registerInteraction(action);

    // Add to ThingsManager
    const things = new ThingsManager();
    things.addThing(thing);

    // Encode
    const openapi = OpenAPIEncoder.encode(things);
    expect(openapi).toMatchSnapshot();
});

test('encode multiple things', () => {
    // Prepare test things
    const thing1 = new Thing('Testthing', 'Thing', 'http://localhost');

    const property1 = new Property('testProperty', new DataSchema('integer', true), true, false);
    property1.registerLink(new HTTPLink('/testproperty', thing1.base, 'application/json'));
    thing1.registerInteraction(property1);

    const action1 = new Action('testAction', new InputSchema('boolean'), new OutputSchema('boolean'));
    action1.registerLink(new HTTPLink('/testaction', thing1.base, 'application/json'));
    thing1.registerInteraction(action1);

    const thing2 = new Thing('Testthing 2', 'Thing');

    const property2 = new Property('testProperty', new DataSchema('integer', true), true, false);
    property2.registerLink(new HTTPLink('http://link1.com/testproperty', '', 'application/json'));
    property2.registerLink(new HTTPLink('http://link2.com/testproperty', '', 'application/json'));
    thing2.registerInteraction(property2);

    // Add to ThingsManager
    const things = new ThingsManager();
    things.addThing(thing1);
    things.addThing(thing2);

    // Encode
    const openapi = OpenAPIEncoder.encode(things);
    expect(openapi).toMatchSnapshot();
});
