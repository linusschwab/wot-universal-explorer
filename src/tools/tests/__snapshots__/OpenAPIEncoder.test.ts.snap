// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`encode multiple things 1`] = `
"{
    \\"openapi\\": \\"3.0.0\\",
    \\"info\\": {
        \\"version\\": \\"1.0.0\\",
        \\"title\\": \\"Universal Explorer for the Web of Things\\",
        \\"description\\": \\"API to interact with things\\"
    },
    \\"servers\\": [
        {
            \\"url\\": \\"http://localhost:5000/things\\",
            \\"description\\": \\"Development server\\"
        }
    ],
    \\"paths\\": {
        \\"/testthing/properties/testproperty\\": {
            \\"get\\": {
                \\"description\\": \\"Read Property testProperty\\",
                \\"responses\\": {
                    \\"200\\": {
                        \\"description\\": \\"Ok\\",
                        \\"content\\": {
                            \\"application/json\\": {
                                \\"schema\\": {
                                    \\"type\\": \\"integer\\"
                                }
                            }
                        }
                    },
                    \\"400\\": {
                        \\"description\\": \\"Bad request\\"
                    },
                    \\"408\\": {
                        \\"description\\": \\"Remote thing did not respond\\"
                    }
                },
                \\"tags\\": [
                    \\"Testthing\\"
                ]
            },
            \\"put\\": {
                \\"description\\": \\"Write Property testProperty\\",
                \\"responses\\": {
                    \\"200\\": {
                        \\"description\\": \\"Ok\\",
                        \\"content\\": {
                            \\"application/json\\": {
                                \\"schema\\": {
                                    \\"type\\": \\"integer\\"
                                }
                            }
                        }
                    },
                    \\"400\\": {
                        \\"description\\": \\"Bad request\\"
                    },
                    \\"408\\": {
                        \\"description\\": \\"Remote thing did not respond\\"
                    }
                },
                \\"tags\\": [
                    \\"Testthing\\"
                ],
                \\"requestBody\\": {
                    \\"description\\": \\"Property testProperty parameters\\",
                    \\"required\\": \\"true\\",
                    \\"content\\": {
                        \\"application/json\\": {
                            \\"schema\\": {
                                \\"type\\": \\"integer\\"
                            }
                        }
                    }
                }
            }
        },
        \\"/testthing/actions/testaction\\": {
            \\"post\\": {
                \\"description\\": \\"Invoke Action testAction\\",
                \\"responses\\": {
                    \\"200\\": {
                        \\"description\\": \\"Ok\\",
                        \\"content\\": {
                            \\"application/json\\": {
                                \\"schema\\": {
                                    \\"type\\": \\"boolean\\"
                                }
                            }
                        }
                    },
                    \\"400\\": {
                        \\"description\\": \\"Bad request\\"
                    },
                    \\"408\\": {
                        \\"description\\": \\"Remote thing did not respond\\"
                    }
                },
                \\"tags\\": [
                    \\"Testthing\\"
                ],
                \\"requestBody\\": {
                    \\"description\\": \\"Action testAction parameters\\",
                    \\"required\\": \\"true\\",
                    \\"content\\": {
                        \\"application/json\\": {
                            \\"schema\\": {
                                \\"type\\": \\"boolean\\"
                            }
                        }
                    }
                }
            }
        },
        \\"/testthing-2/properties/testproperty\\": {
            \\"get\\": {
                \\"description\\": \\"Read Property testProperty\\",
                \\"responses\\": {
                    \\"200\\": {
                        \\"description\\": \\"Ok\\",
                        \\"content\\": {
                            \\"application/json\\": {
                                \\"schema\\": {
                                    \\"type\\": \\"integer\\"
                                }
                            }
                        }
                    },
                    \\"400\\": {
                        \\"description\\": \\"Bad request\\"
                    },
                    \\"408\\": {
                        \\"description\\": \\"Remote thing did not respond\\"
                    }
                },
                \\"tags\\": [
                    \\"Testthing 2\\"
                ]
            },
            \\"put\\": {
                \\"description\\": \\"Write Property testProperty\\",
                \\"responses\\": {
                    \\"200\\": {
                        \\"description\\": \\"Ok\\",
                        \\"content\\": {
                            \\"application/json\\": {
                                \\"schema\\": {
                                    \\"type\\": \\"integer\\"
                                }
                            }
                        }
                    },
                    \\"400\\": {
                        \\"description\\": \\"Bad request\\"
                    },
                    \\"408\\": {
                        \\"description\\": \\"Remote thing did not respond\\"
                    }
                },
                \\"tags\\": [
                    \\"Testthing 2\\"
                ],
                \\"requestBody\\": {
                    \\"description\\": \\"Property testProperty parameters\\",
                    \\"required\\": \\"true\\",
                    \\"content\\": {
                        \\"application/json\\": {
                            \\"schema\\": {
                                \\"type\\": \\"integer\\"
                            }
                        }
                    }
                }
            }
        }
    }
}"
`;

exports[`encode one thing 1`] = `
"{
    \\"openapi\\": \\"3.0.0\\",
    \\"info\\": {
        \\"version\\": \\"1.0.0\\",
        \\"title\\": \\"Universal Explorer for the Web of Things\\",
        \\"description\\": \\"API to interact with things\\"
    },
    \\"servers\\": [
        {
            \\"url\\": \\"http://localhost:5000/things\\",
            \\"description\\": \\"Development server\\"
        }
    ],
    \\"paths\\": {
        \\"/testthing/properties/testproperty\\": {
            \\"get\\": {
                \\"description\\": \\"Read Property testProperty\\",
                \\"responses\\": {
                    \\"200\\": {
                        \\"description\\": \\"Ok\\",
                        \\"content\\": {
                            \\"application/json\\": {
                                \\"schema\\": {
                                    \\"type\\": \\"integer\\"
                                }
                            }
                        }
                    },
                    \\"400\\": {
                        \\"description\\": \\"Bad request\\"
                    },
                    \\"408\\": {
                        \\"description\\": \\"Remote thing did not respond\\"
                    }
                },
                \\"tags\\": [
                    \\"Testthing\\"
                ]
            },
            \\"put\\": {
                \\"description\\": \\"Write Property testProperty\\",
                \\"responses\\": {
                    \\"200\\": {
                        \\"description\\": \\"Ok\\",
                        \\"content\\": {
                            \\"application/json\\": {
                                \\"schema\\": {
                                    \\"type\\": \\"integer\\"
                                }
                            }
                        }
                    },
                    \\"400\\": {
                        \\"description\\": \\"Bad request\\"
                    },
                    \\"408\\": {
                        \\"description\\": \\"Remote thing did not respond\\"
                    }
                },
                \\"tags\\": [
                    \\"Testthing\\"
                ],
                \\"requestBody\\": {
                    \\"description\\": \\"Property testProperty parameters\\",
                    \\"required\\": \\"true\\",
                    \\"content\\": {
                        \\"application/json\\": {
                            \\"schema\\": {
                                \\"type\\": \\"integer\\"
                            }
                        }
                    }
                }
            }
        },
        \\"/testthing/actions/testaction\\": {
            \\"post\\": {
                \\"description\\": \\"Invoke Action testAction\\",
                \\"responses\\": {
                    \\"200\\": {
                        \\"description\\": \\"Ok\\",
                        \\"content\\": {
                            \\"application/json\\": {
                                \\"schema\\": {
                                    \\"type\\": \\"boolean\\"
                                }
                            }
                        }
                    },
                    \\"400\\": {
                        \\"description\\": \\"Bad request\\"
                    },
                    \\"408\\": {
                        \\"description\\": \\"Remote thing did not respond\\"
                    }
                },
                \\"tags\\": [
                    \\"Testthing\\"
                ],
                \\"requestBody\\": {
                    \\"description\\": \\"Action testAction parameters\\",
                    \\"required\\": \\"true\\",
                    \\"content\\": {
                        \\"application/json\\": {
                            \\"schema\\": {
                                \\"type\\": \\"boolean\\"
                            }
                        }
                    }
                }
            }
        }
    }
}"
`;
